Add mappings in generated Rust code to eliminate some boilerplates and improve safety

emgre opened the pull request:
What was supposed to just be a check of the enums to bypass the undefined behaviour raised in #23 ended up in a large refactoring of the Rust codegen with more powerful mappings to eliminate as much manual labour as possible.

Because some mappings can happen in structs, the solution was to hide the C fields and provide public accessors/mutators to interact with them in a Rusty fashion. To help creating new structs from scratch, a `MyStructFields` (notice the extra "Fields") is generated for each `MyStruct`. This struct has all its Rusty members public and implements `From<MyStructFields> for MyStruct` meaning you can do something like:

```rust
let example: MyStruct = MyStructField {
   foo: MyEnum::bar,
   // ...
}.into();
```

The current mappings are:
- Enums are mapped from `c_int` to `MyEnum`. The conversion makes sure that the enum value is valid and panics if it's not. This is better than the undefined behaviour of current Rust.
- StructRefs are mapped from `*const MyStruct` to `Option<&MyStruct>`. `None` is mapped to `std::ptr::null()`.
- Durations are mapped from `u64`/`f64` to `std::time::Duration`.
- Strings are mapped from `*const c_char` to `&std::ffi::CStr`.

Also, I removed the empty safety comments in the generated code, so we'll need to add a `#![allow(clippy::missing_safety_doc)]` at the beginning of our `ffi.rs` files. For some reason, I cannot put this in the generated file.

jadamcrain commented:
Those kind of crate-level annotations may only be put at the top of `lib.rs`.

emgre commented:
They can be put on top of any module. The problem seems to be that we use `include!()` to include the generated file, so it's not exactly at the beginning of a module file.

jadamcrain commented:
Can we add a function level annotation to just ignore it?

emgre commented:
I added the allow line on every function now. I also added a `#[allow(clippy::needless_lifetime)]`, because it gets just too complicated to handle the lifetime printing.
